### PR TITLE
Workflows

### PR DESCRIPTION
**Pull request by nateshim-indico**
Main changes are `IndicoClientWrapper`, `Workflows`, and their respective test files.

### CODE REVIEW

File: IndicoToolkit.Tests/AutoClass/AutoClassifierTests.cs [hunk 1]
```diff
@@ -68,20 +70,22 @@ public async Task TestToCSV()
...
-                    } else {
+                        shouldBeTwoAclass++;
+                    }
+                    else
+                    {
                         Assert.Equal("b", rec.docClass);
                         Assert.DoesNotContain("YY", rec.docText);
                         Assert.DoesNotContain("ZZ", rec.docText);
-                        shouldBeOneBclass ++;
+                        shouldBeOneBclass++;
```
Comment (Scott771) [inline IndicoToolkit.Tests/AutoClass/AutoClassifierTests.cs:88]: should we be asserting what the count of these are at the end? Assert.Equal(1, shouldBeOneBclass)?

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using IndicoV2;
+using IndicoV2.Storage.Models;
+using IndicoV2.Jobs.Models;
+
+using IndicoToolkit.Types;
+using IndicoToolkit.Exception;
+
+namespace IndicoToolkit.IndicoWrapper
```
Comment (Scott771) [inline IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:13]: It's probably not worth it here, but I hate the name "IndicoWrapper"-- feel same way for the python toolkit...

Comment (Scott771) [reply IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:13]: same for "indicoClientWrapper" obviously...

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        public async Task<List<Prediction>> GetPredictionsWithModelId(int modelId, List<string> samples)
+        {
+            List<Prediction> predictions = new List<Prediction>();
+            string jobId = await client.Models().Predict(modelId, samples, default);
+            dynamic jobResults = await client.Jobs().GetResultAsync<dynamic>(jobId);
+            JobStatus status = await GetJobStatus(jobId);
+            if (status != JobStatus.SUCCESS)
+            {
+                throw new ToolkitStatusException(
+                    $"Predictions Failed, {status}"
```
Comment (Scott771) [inline IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:89]: does "status" give us anything other than a string of "Failed" i.e. would that give us an error message?

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        /// </summary>
+        /// <param name="modelId">The model Id to submit to</param>
+        /// <param name="samples">A list containing the text samples you want to submit</param>
+        /// <returns>A list of Predictions</returns>
+        public async Task<List<Prediction>> GetPredictionsWithModelId(int modelId, List<string> samples)
+        {
+            List<Prediction> predictions = new List<Prediction>();
+            string jobId = await client.Models().Predict(modelId, samples, default);
+            dynamic jobResults = await client.Jobs().GetResultAsync<dynamic>(jobId);
+            JobStatus status = await GetJobStatus(jobId);
```
Comment (Scott771) [inline IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:85]: should this switch lines with 84? I.e. should we check the status before we retrieve the results?

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs (hunk 2 above)
Comment (nateshim-indico) [reply IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:13]: Agreed. Wonder what your thoughts are on this (and I don't feel too strongly on this), but I think the idea of having this `indico_wrapper.py` as a general hubspot of functionality within the toolkit can be confusing, given that the toolkit itself is more like an assortment of modules. Could be worthwhile branching out indicowrapper fully, much like what we have for Datasets and Workflow module.

File: IndicoToolkit/IndicoWrapper/Workflows.cs [hunk 5]
```diff
@@ -1,13 +1,201 @@
...
-        public IndicoClient client;
-        public Workflows(IndicoClient _client)
+        public Workflows(IndicoClient _client) : base(_client) { }
+
+
+        /// <summary>
+        /// Retrieves workflow with given datasetID
+        /// </summary>
+        /// <param name="datasetId">ID of corresponding dataset</param>
+        public async Task<IEnumerable<IWorkflow>> GetWorkflow(int datasetId)
```
Comment (Scott771) [inline IndicoToolkit/IndicoWrapper/Workflows.cs:28]: do we need this call? just thinking that this call might be antiquated... i.e. multiple workflows can be associated with one datasetId

File: IndicoToolkit.Tests/AutoClass/AutoClassifierTests.cs (hunk 1 above)
Comment (nateshim-indico) [reply IndicoToolkit.Tests/AutoClass/AutoClassifierTests.cs:88]: line 92 and 93 should be doing that, although I can move them inside line 69

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs (hunk 3 above)
Comment (nateshim-indico) [reply IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:89]: ToolkitStatusException inherits from System.Exception and calls base(message) so yes, the entire thread should be reflected

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs (hunk 4 above)
Comment (nateshim-indico) [reply IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:85]: oh yeah that makes sense

File: IndicoToolkit/IndicoWrapper/Workflows.cs (hunk 5 above)
Comment (nateshim-indico) [reply IndicoToolkit/IndicoWrapper/Workflows.cs:28]: makes sense

File: IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs (hunk 2 above)
Comment (Scott771) [reply IndicoToolkit/IndicoWrapper/IndicoClientWrapper.cs:13]: I'm totally down-- let's go for it

File: IndicoToolkit/IndicoWrapper/Workflows.cs [hunk 6]
```diff
@@ -1,13 +1,191 @@
...
+        /// <param name="returnRawJson">If true return raw json result, otherwise return WorkflowResult object</param>
+        /// <param name="throwExceptionForFailed">if true, ToolkitStatusError raised for failed submissions</param>
+        /// <param name="returnFailedResults">if true, return objects for failed submissions</param>
+        /// <param name="ignoreDeletedSubmissions">if true, ignore deleted submissions</param>
+        /// <returns>workflow result objects</returns>
+        public async Task<List<dynamic>> GetSubmissionResultsFromIds
+        (
+            List<int> submissionIds,
+            int timeout = 180,
+            bool returnRawJson = false,
```
Comment (Scott771) [inline IndicoToolkit/IndicoWrapper/Workflows.cs:104]: remove rawjson kwarg and change output type to be wflowobject

File: IndicoToolkit/IndicoWrapper/Workflows.cs [hunk 7]
```diff
@@ -1,13 +1,191 @@
...
+            ISubmission job = await client.Submissions().GetAsync(submissionId);
+            return await GetStorageObject(job.ResultFile);
+        }
+
+        /// <summary>
+        /// Submits submission to review
+        /// </summary>
+        /// <param name="submissionId">Id of the corresponding submission</param>
+        /// <returns>JobId of submission</returns>
+        public async Task<JObject> SubmitSubmissionReview(int submissionId)
```
Comment (Scott771) [inline IndicoToolkit/IndicoWrapper/Workflows.cs:159]: remove this for now or figure out why it exists

**Review by Scott771 — APPROVED** [review]

**Pull request merged**